src/raytracer.rs

```diff
@@ -1,90 +1,76 @@
-extern crate num_cpus;
-extern crate crossbeam_utils;
-
-use na::Vector3 as Vec3;
-use std::f64::INFINITY;
 use camera::Camera;
 use light::Light;
 use object::Object;
-use std::sync::Arc;
+use surface::{Division, Surface};
+
+use na::Vector3 as Vec3;
+use scoped_threadpool::Pool;
+use num_cpus;
+
+use std::f64::INFINITY;
 
 pub struct Ray {
     pub dir: Vec3<f64>,
     pub origin: Vec3<f64>
 }
 
 pub fn march (cam: &Camera, objects: &Vec<Box<Object>>, lights: &Vec<Box<Light>>)
-              -> Vec<f64> {
-    let pixel_num = (cam.height * cam.width * 3) as usize;
-    let mut pixels: Vec<f64> = Vec::with_capacity(pixel_num);
-    unsafe {
-        pixels.set_len(pixel_num);
-    }
-    let num_cpus = num_cpus::get();
-    let t_cam = Arc::new((*cam).clone());
-    let t_obj = Arc::new((*objects).clone());
-    let t_lights = Arc::new((*lights).clone());
-    let mut chunk_num = 0;
-    for chunk in pixels.chunks_mut(pixel_num / num_cpus) {
-        let w_cam = t_cam.clone();
-        let w_obj = t_obj.clone();
-        let w_lights = t_lights.clone();
-        let w_num_cpus = num_cpus;
-        let w_i = chunk_num;
-        crossbeam_utils::scoped::scope(|scope| {
-            scope.spawn(move || {
-                process_part(w_cam, w_obj, w_lights,
-                             w_num_cpus, w_i, chunk);
+              -> Vec<u8> {
+
+    let mut pool = Pool::new(num_cpus::get() as u32);
+    let surf = Surface::new(cam.width, cam.height);
+
+    pool.scoped(|scope| {
+        for chunk in surf.divide(32, 32) {
+            scope.execute(move || {
+                process_part(cam, objects, lights, chunk);
             });
-        });
-        chunk_num += 1;
-    }
-    pixels
+        }
+    });
+    surf.pixels
 }
 
-fn process_part(cam: Arc<Camera>, objects: Arc<Vec<Box<Object>>>, lights: Arc<Vec<Box<Light>>>,
-                num_cpus: usize, part_num: usize, pixels: &mut [f64]) {
+fn process_part(cam: &Camera, objects: &Vec<Box<Object>>, lights: &Vec<Box<Light>>,
+                chunk: Division) {
     let aspect: f64 = cam.width as f64 / cam.height as f64;
```

Comment:
Casting floats may be lossy if you the size of float from 64, maybe use `f64::from` (there is more of this below too)

Reply:
Hmm, not sure that I should consider that in this case.